src/main/java/org/snakeyaml/engine/v2/composer/Composer.kt

```diff
@@ -0,0 +1,359 @@
+     *
+     * If the stream contains more than one document an exception is thrown.
+     *
+     * @return The root node of the document or `Optional.empty()` if no document is
+     * available.
+     */
+    val singleNode: Optional<Node>
+        get() {
+            // Drop the STREAM-START event.
+            parser.next()
+            // Compose a document if the stream is not empty.
+            val document = if (!parser.checkEvent(Event.ID.StreamEnd)) {
+                Optional.of(next())
+            } else {
+                Optional.empty<Node>()
+            }
+            // Ensure that the stream contains no more documents.
+            if (!parser.checkEvent(Event.ID.StreamEnd)) {
+                val event = parser.next()
+                val previousDocMark = document.flatMap { obj: Node -> obj.startMark }
+                throw ComposerException(
+                    problem = "expected a single document in the stream",
+                    problemMark = previousDocMark,
+                    context = "but found another document",
+                    contextMark = event.startMark,
+                )
+            }
+            // Drop the STREAM-END event.
+            parser.next()
+            return document
+        }
+
+    /**
+     * Reads and composes the next document.
+     *
+     * @return The root node of the document or `null` if no more documents are available.
+     */
+    override fun next(): Node {
+        // Collect inter-document start comments
+        blockCommentsCollector.collectEvents()
+        if (parser.checkEvent(Event.ID.StreamEnd)) {
+            val commentLines = blockCommentsCollector.consume()
+            val startMark = commentLines[0].startMark
+            val children = mutableListOf<NodeTuple>()
+            val node: Node = MappingNode(
+                tag = Tag.COMMENT,
+                resolved = false,
+                value = children,
+                flowStyle = FlowStyle.BLOCK,
+                startMark = startMark,
+                endMark = Optional.empty(),
+            )
+            node.blockComments = (commentLines)
+            return node
+        }
+        // Drop the DOCUMENT-START event.
+        parser.next()
+        // Compose the root node.
+        val node = composeNode(Optional.empty())
+        // Drop the DOCUMENT-END event.
+        blockCommentsCollector.collectEvents()
+        if (!blockCommentsCollector.isEmpty()) {
+            node.endComments = blockCommentsCollector.consume()
+        }
+        parser.next()
+        anchors.clear()
+        recursiveNodes.clear()
+        nonScalarAliasesCount = 0
+        return node
+    }
+
+    private fun composeNode(parent: Optional<Node>): Node {
+        blockCommentsCollector.collectEvents()
+        parent.ifPresent { e: Node ->
+            recursiveNodes.add(e) // TODO add unit test for this line
+        }
+        val node: Node
+        if (parser.checkEvent(Event.ID.Alias)) {
```

Comment:
This type-checking looks like a good candidate for making `Event` a sealed class and using a 'when' statement.